Discard updatebattleinfo for not connected users

FIr3baL opened the pull request:
Hi,
We faced a problem on bar infrastructure, where chobby shows players in battle, which were already removed from battle and are disconnected from server.
This is caused by receiving CLIENTBATTLESTATUS after LEFTBATTLE and REMOVEUSER was already sent. This leads to adding the user again to battle.

With this PR I added a check, if user is in list of connected users before adding him to the battle by CLIENTBATTLESTATUS.

Only testet in bar infrastructure. Can´t tell if it conflicts with other protocols and infrastructures.
Hope chances are good, that this has no negative effect in other environments.

best regards
Fireball

**Added: Don´t check user connected if updating status of a bot**

FIr3baL commented:
Don´t merge yet.
Saw a problem when adding bots by chobby.

FIr3baL commented:
Imho, this can be merged now.
The mentioned issue with bots doesn´t belong to this change.

gajop commented on libs/liblobby/lobby/interface.lua:
```diff
@@ -191,6 +191,54 @@ local function UpdateAndCreateMerge(userData, status)
 	return battleStatus, updated
 end
 
+--n = pow(2,i) -- where i = 0,31
+--print('{',n//1000000,',', n%1000000,'},')
+local bin2decmillion16 = { -- the <1M and >1M parts of 2^nth powers where n > 16
+	{ 0 , 65536 }, --16
+	{ 0 , 131072 }, --17
+	{ 0 , 262144 }, --18
+	{ 0 , 524288 }, --19
+	{ 1 , 48576 },
+	{ 2 , 97152 },
+	{ 4 , 194304 },
+	{ 8 , 388608 },
+	{ 16 , 777216 }, --24
+	{ 33 , 554432 }, --25
+	{ 67 , 108864 }, --26
+	{ 134 , 217728 }, --27
+	{ 268 , 435456 }, --28
+	{ 536 , 870912 },  --29
+	{ 1073 , 741824 }, --30
+	{ 2147 , 483648 }, -- 31
+	}
+
+-- Combine two 16 bit numbers into a string-formatted 32-bit integer
+local function lsbmsb16tostring(lsb,msb)
```
I'd like this code to be unit tested. I don't trust human reviewers/authors to be correct here (as history has shown).

I'm not sure if current Chobby has unit testing framework setup, but luaunit is popular and not that difficult to add. I've used it myself to test a promise library I wrote, see https://github.com/gajop/spring-lua-utils/blob/c1ed690f660c8f46edd3a35f6bceca19eab2c79b/promise.lua

On code style, I think this doesn't match Chobby's syntax. I believe Chobby uses camelCase for variables and this is all lowercase.

gajop commented on libs/liblobby/lobby/lobby.lua:
```diff
@@ -898,6 +898,11 @@ end
 -- Bots have isBot=true, AIs have aiLib~=nil and humans are the remaining
 -- Example: _OnUpdateUserStatus("gajop", {isAway=false, isInGame=true})
 function Lobby:_OnUpdateUserBattleStatus(userName, status)
+	if status.owner == nil and not self.users[userName] or
+	   status.owner and not self.users[status.owner] then
+		Spring.Echo("[liblobby] Error OnUpdateUserBattleStatus: Tried to update non connected user in battle: ", userName)
+		return
+	end
```
This change (the main change of this PR?) seems unrelated and should be done in a different PR.

I also the conditions in this if are confusing. Maybe you meant to use parentheses here?

```lua
(status.owner == nil and not self.users[userName]) or
(status.owner ~= nil and not self.users[status.owner])
```

gajop commented on libs/liblobby/lobby/lobby.lua:
```diff
@@ -898,6 +898,11 @@ end
 -- Bots have isBot=true, AIs have aiLib~=nil and humans are the remaining
```
Docs above this line don't even mention `status.owner`. Is this a key that exists? If so, docs should be updated - or maybe removed if they're confusing.

That said, I can see `status.owner` being used below so it probably does exist and I can understand if you don't want to do it now - it's not a part of your PR technically.

gajop commented on libs/liblobby/lobby/lobby.lua:
```diff
@@ -898,6 +898,11 @@ end
 function Lobby:_OnUpdateUserBattleStatus(userName, status)
+	if status.owner == nil and not self.users[userName] or
+	   status.owner and not self.users[status.owner] then
+		Spring.Echo("[liblobby] Error OnUpdateUserBattleStatus: Tried to update non connected user in battle: ", userName)
```
I recommend using Spring.Log so Spring correctly formats this.

This should work I think:
```lua
Spring.Log("liblobby", LOG.ERROR, "Error OnUpdateUserBattleStatus: Tried to update non connected user in battle: ", userName)
```

gajop reviewed (commented):
Here's my brief look at it - I recommend splitting PRs so it's easier to merge.
I didn't look into the implementation details for bit ops, I think that should just be unit tested instead (which is its own hurdle I think - so maybe can be done in a few PRs).

FIr3baL commented:
@gajop Wow, thx for this nice review.
I´ll split the PR and you are right with the parentheses. Will add some docu to the function about owner as well !

FIr3baL commented:
- isolated this PR from change in Encode/Decode BattleStatus
- changed log to Spring.LOG
- docs of function _OnUpdateUserBattleStatus seemed to be lazy copy of _OnUpdateUserStatus docs - updated to correct keys
- added parentheses to if conditions

+ found a wrong usage of status.playMode and replaced by status.isSpectator
